File: String_sol/string to integer(atoi)/string to integer(atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        # Edge case: empty string
        if len(s) == 0:
            return 0

        i = 0
        n = len(s)

        # 1. Skip leading spaces
        while i < n and s[i] == ' ':
            i += 1

        # Cut off leading spaces
        s = s[i:]
        if not s:   # If string became empty after trimming spaces
            return 0

        # 2. Handle sign
        sign = 1
        if s[0] == '-':
            sign = -1

        # Integer range (32-bit signed)
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31

        # 3. Skip sign char if present
        i = 1 if s[0] in ['+', '-'] else 0

        ans = 0

        # 4. Read digits
        while i < len(s):
            if not s[i].isdigit():   # Stop if non-digit encountered
                break

            # Convert char to int (e.g., '7' → 7)
            digit = ord(s[i]) - ord('0')

            ans = ans * 10 + digit   # Build number

            # 5. Check for overflow/underflow
            if sign == -1 and -ans < INT_MIN:
                return INT_MIN
            if sign == 1 and ans > INT_MAX:
                return INT_MAX

            i += 1

        # 6. Apply sign and return
        return sign * ans

File: String_sol/string to integer(atoi)/string to integer(atoi).cpp
#include<iostream>
#include<string>
#include<climits>

using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        int i = 0;
        int n = s.size();
        int sign = 1;
        long long num = 0;

        while(i < n && s[i] == ' ') i++;

        if(i<n && s[i] == '+' || s[i] == '-') {
            if(s[i]=='-') sign = -1;
            i++;
        }

        while(i<n && isdigit(s[i])) {
            int digit = s[i] - '0';
            num = num*10+digit;

            if(sign*num>INT_MAX) return INT_MAX;
            if(sign*num<INT_MIN) return INT_MIN;

            i++;
        }

        return sign*num;
    }
};


// most optimal approach of the code.



class Solution {
public:
    int myAtoi(string s) {
        if(s.length() == 0) return 0;

        int i = 0;
        while(i<s.size() && s[i] == ' ') i++;
        s = s.substr(i);

        int sign = +1;
        long long ans = 0;

        if(s[0] == '-') sign= -1;

        int MAX = INT_MAX, MIN = INT_MIN;
        i=(s[0] == '+' || s[0] == '-')? 1 : 0;

        while(i<s.length()) {
            if(s[0] == ' ' || !isdigit(s[i])) break;

            ans = ans*10 + s[i] - '0';
            if(sign == -1 && -1*ans<MIN) return MIN;
            if(sign == 1 && ans>MAX) return MAX;

            i++;
        }
        return (int)(sign*ans);
    }
};